### ISSUE TITLE
Expose Collectors for Multisets accepting a count function

### ISSUE THREAD

**Issue by lowasser** [google/guava#2696]
This is the main issue I perceive as outstanding originally raised from https://github.com/google/guava/issues/2694.

**Comment by perceptron8** [comment]
Indeed. Give me some time, please. I need to collect my thoughts.

**Comment by lowasser** [comment]
Here are mine, FWIW:

I see the appropriate API here as 

    ImmutableMultiset.toImmutableMultiset(Function<T, E>, ToIntFunction<T>)
    ImmutableSortedMultiset.toImmutableSortedMultiset(
        Comparator<E>, Function<T, E>, ToIntFunction<T>)
    Multisets.toMultiset(Function<T, E>, ToIntFunction<T>, Supplier<Multiset>)

This seems like the appropriate subset: I don't think we should have Collectors for each mutable Multiset implementation when the Supplier approach works sufficiently, and it seems perfectly acceptable to depend on Collectors.toCollection for the `e -> 1` case.  The above seems to me like the right amount of API surface area.  This will need to go through the internal API review process, though.

**Comment by perceptron8** [comment]
You are right that keyFunction is equally useful as countFunction. I was concerned about the similarity of
* `toImmutableMap(Function keyFunction, Function valueFunction)` to
* `toImmutableMultiset(Function elementFunction, ToIntFunction countFunction)`.

The former one throws if there is at least one duplicate key mapping. The latter uses implicit, although obvious "mergeFunction" (addAll). This is fully described in javadoc, which I missed. No more confusion probably.

`Multisets.toMultiset` with `Supplier` looks good too (as good as already existing `Multimaps.toMultimap`).

I perceive no natural-order collectors as a design decision as every ImmutableSorted* collector requires explicit Comparator. That's consistent and clear. I can live with `naturalOrder()`.

In short, proposed API seems appropriate.

I keep my fingers crossed for a positive result of the review! ;)

**Comment by perceptron8** [comment]
Almost there? See https://github.com/google/guava/commit/6df8b3ca1d1d6b2dfe3aff0316079da61937eef9#commitcomment-20466808 - and don't kill me, please.

It's very nice that you decided to expose these Collectors. Thank you again!

**Comment by perceptron8** [comment]
@lowasser Could you, please, check above? It's just about renaming (and consistency). I can't wait to mark this as resolved.

**Comment by perceptron8** [comment]
@lowasser I think this can be closed because of 15c2b77.

**Comment by perceptron8** [comment]
@lowasser I think that there is still one thing missing: wildcards. I came to this today, when I tried to replace my own implementation with delegation to `Multisets.toMultiset`. I wish I had noticed this before.

Instead of
```
public static <T, E, M extends Multiset<E>> Collector<T, ?, M> toMultiset(
      Function<T, E> elementFunction,
      ToIntFunction<T> countFunction,
      Supplier<M> multisetSupplier
)
```

one would expect
```
public static <T, E, M extends Multiset<E>> Collector<T, ?, M> toMultiset(
	Function<? super T, ? extends E> elementFunction,
	ToIntFunction<? super T> countFunction,
	Supplier<M> multisetSupplier
)
```

See `Mutlimaps.toMultimap` which is similar but correct.

**Comment by cpovirk** [comment]
We have a standing team disagreement about the importance of `? extends E`, but I think we all agree on the need for `? super T` :) I'm sending a CL.

**Comment by perceptron8** [comment]
@cpovirk Well, it's still inconsistent with [ImmutableMultiset.toImmutableMultiset(Function, ToIntFunction)](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/collect/ImmutableMultiset.html#toImmutableMultiset-java.util.function.Function-java.util.function.ToIntFunction-). Does it have to be? Please, please, just use PECS. It is not only more flexible but also more idiomatic.

**Comment by perceptron8** [comment]
@cpovirk Check [ImmutableSortedMultiset.toImmutableSortedMultiset(Comparator, Function, ToIntFunction)](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/collect/ImmutableSortedMultiset.html#toImmutableSortedMultiset-java.util.Comparator-java.util.function.Function-java.util.function.ToIntFunction-) too!

**Comment by cpovirk** [comment]
This is a little weirder than the vanilla PECS case: It's not about permitting things that wouldn't have been permitted before, just about letting you force a different return type. (Actually, it's possible that the `Collector` case is still weirder than that; I haven't thought about it in detail.) We've spent a surprising amount of time on this discussion without reaching a conclusion, so I think we're just going to leave things as they are (even if it's inconsistent), rather than try again to reach consensus.

**Comment by kevinb9n** [comment]
We apply PECS consistently, but it only discusses types that are produced and consumed, and says nothing about "pass-through types", which you can look at as being neither produced nor consumed. We don't add additional wildcards and type parameters that only serve to massage the return type. The other approach was to decide that such types are *both* produced and consumed, and that would have been okay too, but is not the way we decided, and we would have to add a LOT more wildcards to Guava to do it the other way.

Sometimes we mess up. If there's code that should compile and doesn't, and isn't fixed by just adjusting what return type the code "prefers", then we need to add something. We've also messed up in the opposite direction: no one should ever have had to write `Predicates.<String>alwaysTrue()` because `alwaysTrue` should always have returned `Predicate<Object>` (because that's literally what it is: a predicate that can handle any object).

**Comment by kevinb9n** [comment]
Sorry, I meant to cancel the previous comment, because I remembered that we may have opted for our Collectors to be a special case that match upstream Collectors instead of other Guava APIs. I leave it up to @lowasser what should be done.

**Comment by perceptron8** [comment]
@lowasser Could you please take a look?

**Comment by perceptron8** [comment]
@kevinb9n Any chances before Guava 23?
